### PR TITLE
Match host network when registering a Consul service at the group level

### PR DESCRIPTION
**Pull request by neilmock**
Nomad multi-interface network support introduces a `Ports` struct that is shared across tasks within a group.  This struct holds a mapping between a network group port to a host IP based on the `host_network` directive in the task group network definition.  This PR introduces this bit of state into the Consul service registration code, matching a port label to a `host_network` IP.

The underlying support comes from the initial introduction of a `Ports` field to the `AllocatedSharedResources` of a job allocation in this commit:  https://github.com/hashicorp/nomad/commit/ad8ced3873.  We take that field and pass it down into the Consul service registration layer.

### CODE REVIEW

File: command/agent/consul/client.go [hunk 1]
```diff
@@ -1453,6 +1453,18 @@ func getAddress(addrMode, portLabel string, networks structs.Networks, driverNet
 
 		// Default path: use host ip:port
 		ip, port := networks.Port(portLabel)
+
+		// If a network group port mapping exists for this label, assign the
+		// proper host network IP.
+		if len(ports) > 0 {
```
Comment (jbarnette) [inline command/agent/consul/client.go:1459]: I don't think you need this conditional.

File: command/agent/consul/structs.go [hunk 2]
```diff
@@ -31,6 +31,9 @@ type WorkloadServices struct {
 	// Networks from the task's resources stanza.
 	Networks structs.Networks
 
+	// Port mappings from the task's group stanza. Includes host network.
```
Comment (jbarnette) [inline command/agent/consul/structs.go:34]: A task is inside a group, so maybe "mappings from the group's network stanza"?

**Review by jbarnette — CHANGES_REQUESTED** [review]
@neilmock Ah, so it wasn't wired up at all?

**Comment by neilmock** [comment]
> 
> 
> @neilmock Ah, so it wasn't wired up at all?

I don't see how it could ever work, and this issue leads me to believe it is a known gap they are working on:  https://github.com/hashicorp/nomad/issues/8801

But goddamn a feature roadmap of some kind would be helpful.

File: command/agent/consul/client.go (hunk 1 above)
Comment (neilmock) [reply command/agent/consul/client.go:1459]: Indeed, will fix.

File: command/agent/consul/structs.go (hunk 2 above)
Comment (neilmock) [reply command/agent/consul/structs.go:34]: Yep, Nomad nomenclature gets me sometimes as my brain always wants to think of jobs as "tasks".  Will rephrase.

**Review by jbarnette — APPROVED** [review]

**Comment by jbarnette** [comment]
👍 assuming you've tried this out in prod.

**Comment by neilmock** [comment]
> 
> 
> 👍 assuming you've tried this out in prod.

I rem'd `-dev` from my local Nomad setup, specified my Packet dev pub/priv IPs as diff host networks in configuration and it worked. Given this incontrovertible evidence of success, I'll merge this, tag, build a new Nomad release, bump `distro` and re-build the infra from scratch.  Cool?

**Pull request merged**